ls1intum/Artemis
# [Bugfix] Illegal submissions grace period

### Checklist


### Motivation and Context


The actual implementation of illegal submission does not take the grace period into account which may lead to false positives when students submit exercises in the exam.

### Description

This pr:
- Now takes the grace period into account and adapts related tests.
- Reverts the filtering of illegal submission in various areas notably the student exam summary, participation, and submission tables.
- Adds database index for submission type
- Fixes a bug in the exam summary view where it crashes when a participation wasn't specified
- Adds a warning to the student's code editor if the submission is illegal

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new exam and make sure you input a large grace period (e.g 999)
4. Add a programming exercise into the exam, participate as a student, and submit once.
3. Do not hand in early but navigate to the student's repository in Bitbucket
4. Find a code file and edit it. This will trigger a new legal submission since it's within the grace period.
5. Hand in the exam and end it.
6. Now navigate again to Bitbucket and edit a file. This should generate an illegal submission since the exam has ended.
7. Navigate to the submission table of the exam and make sure the submission count and types are correct.

### Test Coverage



ProgrammingExerciseStudentParticipationRepository.java 68%
StudentParticipationRepository.java 90%
ExamSubmissionService.java 98%
ProgrammingSubmissionService.java 76%
RepositoryProgrammingExerciseParticipationResource.java 76%

### Screenshots


### Illegal Warning in Exam Summary: 
![image](https://user-images.githubusercontent.com/33484318/114556321-e0fad980-9c68-11eb-9b05-b018652f2f5c.png)

### Participation table shows correct submission count (including illegal)
![image](https://user-images.githubusercontent.com/33484318/114556427-fc65e480-9c68-11eb-9d66-932a0321a510.png)


### Submissions within Grace Period (Last page appears where you submit the exam, do not submit yet and push changes to the repo) still count as `graded`:
#### Submit code while being in this screen: 
![image](https://user-images.githubusercontent.com/33484318/114557144-ac3b5200-9c69-11eb-8898-f9452dd2c28b.png)

![image](https://user-images.githubusercontent.com/33484318/114556591-2919fc00-9c69-11eb-904f-66ebb4fef707.png)

### Submissions after submitting the exam (hand in early and last page submit) lead to illegal submissions:
![image](https://user-images.githubusercontent.com/33484318/114556710-42bb4380-9c69-11eb-96f5-83fce4ba4b6a.png)
